### PR TITLE
agent: prevent very old servers re-joining a cluster with stale data 

### PR DESCRIPTION
**Pull request by loshz**
### Description

This PR introduces the concept of Server Metadata, which is server specific information written to a file stored in the configured data directory. During this initial phase, we only store the last seen timestamp in Unix format:

https://github.com/hashicorp/consul/blob/5103a1a233e95c70d2c5dea207da132e37c4dc64/agent/consul/server_metadata.go#L17-L19

2 new agent methods for attempting to prevent old servers from rejoining an existing cluster. 

- `persistServerMetadata()`: Periodically write a server's metadata to a file in the configured data directory every hour.
- `checkServerLastSeen()`: Attempt to read a server's last seen file and check the Unix timestamp against a configurable max age. If the last seen file does not exist, we treat this as an initial startup and return no error.

#### Example

We attempt to start a previously running server with the following last seen timestamp `1672531200` (2023-Jan-01 00:00:00).

Setting the new config to `server_rejoin_age_max = "3d"` will prevent this server from starting and require an operator to manually remove this file from the data directory if they need to force the server to rejoin.

### Testing & Reproduction steps

- Added unit tests for new Server Metatdata helper funcs.
- Added unit tests for agent Server Metatdata implementation.

### Links

- https://hashicorp.atlassian.net/browse/NET-3772
- https://github.com/hashicorp/consul/blob/main/docs/config/checklist-adding-config-fields.md

### PR Checklist
* [x] Write validation function for new config field.
* [x] updated test coverage
* [x] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

**Review by banks — COMMENTED** [review]
Amazing stuff Dan!

I was originally thinking it might be cleanest to wrap this file handling stuff in a sub-package in the agent package. Then I saw how many actual lines of code it was and was less sure.

But I _think_ I'd still lean towards that because:
 - agent.go is way to big and has all sorts of different logic in. We've been trying to remove stuff from here for years allbeit very slowly.
 - although it's trivial, the logic around reading and writing files, possibly handling misformated (e.g. because of corruption or crash) files etc. would all be so much easier to test in a small package in a unit test vs against a whole agent.

Youd still need the same call points in agent.go so it's only like 20 lines of code to extract but I think it would be a little cleaner that way and a bit easier to test more.

I could also imagine a future where we end up expanding this a bit - e.g. also adding a second timestamp for "last connected to a leader" which might help if servers are up but partitioned for a long time before being decomissioned. I don't think we should do that yet, but it would be much nicer to add stuff like that to a small independent package than agent.go if we do it.

File: agent/agent.go
Comment (boxofrad) [inline agent/agent.go:4602]: I wonder if "older" is a little ambiguous here, and if we could make the error more actionable.

```suggestion
		return fmt.Errorf("refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (%s) - consider wiping your data dir", maxAge)
```

**Review by boxofrad — APPROVED** [review]
Looks great, nice job! 👏🏻 

Don't forget to update the docs. We probably want to call out that you can't set `server_rejoin_age_max` to anything less than an hour because of the frequency we write the file out to disk.

File: agent/config/builder.go
Comment (boxofrad) [inline agent/config/builder.go:1092]: We should validate that it isn't set to anything less than an hour here (because that's how often we write the file out to disk) 💡

Comment (loshz) [reply agent/config/builder.go]: Something like this? https://github.com/hashicorp/consul/pull/17171/commits/21e26a0e01efffcc5fd6b3e2a660dc2a63557558

Comment (boxofrad) [reply agent/config/builder.go]: Nice! Although, maybe this check should be performed in [`builder.validate`](https://github.com/hashicorp/consul/blob/d0beddd9d141868d37b9415a28d78b94fd3d0753/agent/config/builder.go#L1196) - sorry that's my bad, should've been more specific than saying "here" 😅 

We should also add a test case to [`TestLoad_IntegrationWithFlags`](https://github.com/hashicorp/consul/blob/boxofrad/resource-documentation/agent/config/runtime_test.go#L71).

File: agent/agent.go
Comment (jmurret) [inline agent/agent.go:628]: hanging TODO, can we resolve before we merge?

File: agent/agent.go
Comment (jmurret) [inline agent/agent.go:630]: I believe logging loudly was called out in the requirements, can we resolve this TODO by adding that?

File: agent/agent.go
Comment (jmurret) [inline agent/agent.go:4593]: Is it possible that `persistServerMetadata()` never wrote the file because it encountered errors writing the file?  do we need some kind of check between first startup time and now in this so we figure out "it must have had an opportunity to write a file"?  Or possibly an inmemory counter of how many times  `persistServerMetadata()` has run so we should expect a file?

**Review by jmurret — COMMENTED** [review]
Nice work.  I just had a question on whether `persistServerMetadata()` could silently fail many times and then `checkServerLastSeen()` would silently continue because it thought it has an assumption about it just booting up.

**Comment by loshz** [comment]
> Nice work. I just had a question on whether `persistServerMetadata()` could silently fail many times and then `checkServerLastSeen()` would silently continue because it thought it has an assumption about it just booting up.

`checkServerLastSeen()` will always occur first and only once on startup. So even if `persistServerMetadata()` fails multiple times during a previously running server, we wouldn't have a reliable way of telling and therefore shouldn't block a proceeding server startup attempt.

File: agent/agent.go
Comment (loshz) [reply agent/agent.go:4593]: Answered [here](https://github.com/hashicorp/consul/pull/17171#issuecomment-1545476027). But I will update the code comments to explicitly call this out.

File: agent/agent.go
Comment (loshz) [reply agent/agent.go:630]: The requirements state:

>In this state, they log loudly, wait around for 60 seconds and then quit.

I'm not sure what waiting for 60s before exiting achieves as opposed to erroring and exiting immediately. Maybe @banks can chime in on this? I can definitely just add a 60s ticker/loop that logs, if that's the functionality we want to go with.

File: agent/config/builder.go
Comment (loshz) [reply agent/config/builder.go]: Wondering if we want to add a cli flag for this? I'm not sure what the criteria is vs just having a config option only.

File: agent/agent.go
Comment (banks) [reply agent/agent.go:631]: The rationale was two fold:

 1. Just in case someone is not running Consul with a well-behaved supervisor, if we commit immediately it could just end up in a crash loop.
 2. Even if the supervisor is good and throttles it, there is not benefit to crashing quickly since this is a terminal condition. If we just log once and exit it seems much more likely that the operator might have to spend a while looking closely at logs to work out why - there will be one error line saying why surrpoundied by maybe hundreds of lines of shutdoan/startup noise.

So given that it's terminal anyway it seems like a nicer behaviour to force it to be a slow crash loop ourselves - stay up for some time and log more than once so that an operator will immediately see the error line in logs not a see of startup/crash noise.

I'll leave it to you to decide if thats important but it seemed like a better operator UX and should only require a few lines of code or so to loop, log and sleep before finally exiting?

File: agent/config/builder.go
Comment (boxofrad) [reply agent/config/builder.go]: I think we can probably skip the CLI flag for this, as users will very infrequently need to change it.

File: agent/agent.go
Comment (loshz) [reply agent/agent.go]: https://github.com/hashicorp/consul/pull/17171/commits/248055683779771e766382df34d3263d7fc94808

File: website/content/docs/agent/config/config-files.mdx
Comment (banks) [inline website/content/docs/agent/config/config-files.mdx:741]: Would it be clearer not to refer to these internal details and instead say something like:

> If a server is not running for this period, then it will refuse to start up again until an operator intervenes. This is to protect clusters from instability caused by decommissioned servers accidentally being started again.

**Review by banks — APPROVED** [review]
Great job Dan!

Looks good.

I was in two minds about making the file writing slighly more resilient to crashes (i.e. wrote temp file and rename, and/or append a checksum and validate on read).

I came to the conclusion that this is fine - it's possible that the file will be created but empty or filled with junk due to a crash interrupting the write. That means the read will fail on next startup but since we ignore that the worst that happens is that the mechanism is defeated (i.e. same behaviour as before this PR) and that's a somewhat unlikely case since this is protecting from an edge case anyway!

File: website/content/docs/agent/config/config-files.mdx
Comment (loshz) [reply website/content/docs/agent/config/config-files.mdx:739]: I noticed this, too. But did a quick count and to me it seemed more fields didn't have it :sweat_smile:

**Comment by loshz** [comment]
>I came to the conclusion that this is fine - it's possible that the file will be created but empty or filled with junk due to a crash interrupting the write. That means the read will fail on next startup but since we ignore that the worst that happens is that the mechanism is defeated (i.e. same behaviour as before this PR) and that's a somewhat unlikely case since this is protecting from an edge case anyway!

@banks, apologies! I just realised I missed your original comment on this. I agree with the above statement in that the worst that can happen is the behaviour before this change was added.

File: agent/agent.go
Comment (jmurret) [inline agent/agent.go:4572]: Can we resolve this before merging?

**Review by jmurret — APPROVED** [review]
Looks great.  Nice work.  Just left a note about one hanging TODO but approved.  🎉

File: agent/agent.go
Comment (loshz) [reply agent/agent.go:4572]: Whoops - thought I'd done that.

I don't think this is a critical error, as such a failure will not impact behaviour or make the server perform differently.

**Comment by loshz** [comment]
I did some manual testing via Docker compose and everything seems to be working as expected:

Default `server_rejoin_age_max` with a `last_seen_unix` of 2 weeks:
```
tmpaf7jthmgcz-consul-server-1  | ==> Starting Consul agent...
...
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:41:14.616Z [ERROR] agent: startup error: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:41:24.616Z [ERROR] agent: startup error: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:41:34.616Z [ERROR] agent: startup error: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:41:44.617Z [ERROR] agent: startup error: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:41:54.617Z [ERROR] agent: startup error: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:42:04.618Z [ERROR] agent: startup error: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:42:14.620Z [ERROR] agent: Error starting agent: error="refusing to rejoin cluster because server has been offline for more than the configured server_rejoin_age_max (168h0m0s) - consider wiping your data dir"
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T09:42:14.620Z [INFO]  agent: Exit code: code=1
tmpaf7jthmgcz-consul-server-1 exited with code 1
```

Default `server_rejoin_age_max` with a `last_seen_unix` of 10 hours:
```
tmpaf7jthmgcz-consul-server-1  | ==> Starting Consul agent...
...
tmpaf7jthmgcz-consul-server-1  | 2023-05-15T10:40:15.679Z [INFO]  agent: Consul agent running!
```

`server_rejoin_age_max` set to 1 hour:
```
tmpaf7jthmgcz-consul-server-1  | ==> 1 error occurred:
tmpaf7jthmgcz-consul-server-1  | 	* server_rejoin_age_max: duration '1h' cannot be less than: 6h0m0s
tmpaf7jthmgcz-consul-server-1  |
tmpaf7jthmgcz-consul-server-1  |
tmpaf7jthmgcz-consul-server-1 exited with code 1
```

**Comment by markusschuh** [comment]
A feedback too small for a new enhancement issue on its own:

The default of the new parameter [server_rejoin_age_max](https://developer.hashicorp.com/consul/docs/agent/config/config-files#server_rejoin_age_max) is set to

`erver_rejoin_age_max = "168h"`

I'd say, the documentation should use this value - and not translate it to "7d", which is an unsupported format. This is an unnecessary challenge for the less experienced reader - though I agree, it is a small one as the supported syntax is documented at https://developer.hashicorp.com/consul/docs/agent/config/config-files#time-to-live-values

**Comment by david-yu** [comment]
Thanks for the suggestion @markusschuh PR is here: https://github.com/hashicorp/consul/pull/18154

**Pull request merged**
